### DISCUSSION TITLE
Discord Notifications?

### DISCUSSION THREAD

**Discussion by mbround18** [mbround18/valheim-docker#106] [Ideas]
My friends and I were thinking of baking in discord notifications into the server to notify for various events such as restarts & backups. I wanted to get ya'lls opinion, is this a feature you would like available to ya'll?

**Comment by AtroposOrbis** [comment]
Yes please! The people on my server pm me regarding server status quite a bit. A dedicated updates channel would be just lovely! <3

**Reply by TheDakk** [reply]
That would be awesome for my crew as well.

**Reply by mbround18** [reply]
I am thinking of adding this feature after the BepInEx support is added :) There are two methodologies I am thinking about, 
Option A: Building an extension of Odin which would post to a discord webhook when the server is going down or up. 
Option B: Building a separate executable agent which can watch the logs for details and act as a discord bot with commands. 

I am thinking option A right now for simplicities sake.

**Reply by CosmicHorrorDev** [reply]
I think option A would be the cleanest. It would be pretty easy to add that functionality into `send_shutdown()` and start's `invoke`. Scanning the logs also relies on the log messages we're looking for never changing, which seems rather brittle.

**Reply by mbround18** [reply]
Hey @LovecraftianHorror, which rust lib would you suggest for web requests? Reqwest seems chaunky and others seem rather hefty as well. I just dont want to bloat the dependencies by loads.

**Reply by CosmicHorrorDev** [reply]
Hi there :D

If you want a simple one that's quick to compile then there's [ureq](https://lib.rs/crates/ureq). The slow part to compile here would be `rustls`, which is needed for doing HTTPS requests.

And then just to give some alternatives there's also [isahc](https://lib.rs/crates/isahc) (similar dependency-wise and compile time to reqwest. Then likely requiring `libcurl` and `openssl` to be installed, but likely the slimmest dependency and compile time would be [curl](https://lib.rs/crates/curl) (Disclaimer: I never used it, but it seems decently popular and the quickstart makes it look pretty simple).

I think these all do blocking requests by default. If we pulled in `reqwest` then it would be easiest to start out with their blocking api as well (requires a feature flag to expose I think) otherwise we'd have to migrate `odin` to work with async.

All the http clients are pretty similar so it should be relatively easy to switch around and test them out.

Also side-note: after I'm done adding in `odin update` I was planning on doing more work to improve build times, so that cold compilation will be less painful. Once that's up and running, more dependencies is less of an issue on compile times since we should be able to have more incremental compilation. I'll open an issue later today going over what I've found to help with the docker build times!

Sorry for the book lol. There are a lot of options when it comes to web anything in rust

**Reply by J0hnHawk** [reply]
My server is running on my NAS at home. Unfortunately I get a new IP-adress every night. And to bad the server is not everytime listed in game-server-list properly. The server restarts by cronjob after the point a get a new IP-adress. It would be great if the Discord-restart-message contains the (new) IP-adress of the server.

**Reply by CosmicHorrorDev** [reply]
You can send custom notifications with `odin notify` so it should be easy enough for you to script this behavior. Also can you not register a static IP with your ISP? I do know some ISPs restrict or force you to pay for this, really just curious

**Reply by J0hnHawk** [reply]
Unfortunately with my ISP it is not possible to get a static IP for home users :-/

So sending custom notifications would be the easiest way for me or better for my friends... if I wasn't new to Docker and totally new to Odin :-D

Until I figure out how to send these notifications, my friends will still have to do a DNS lookup on my dyndns name ;-)

**Reply by CosmicHorrorDev** [reply]
That's kinda what I figured.

And I hope it's not too rough for you. If you have curl installed then you can get your public ip with

```bash
curl ifconfig.me
```

And you can run one off commands in the docker container using `docker exec` so something like

```bash
docker exec --user steam $CONTAINER_NAME odin notify $(curl ifconfig.me)
```

Should get you started assuming you already have webhooks setup.

Alternatively one of my friends talked about how he used [this service](https://www.noip.com/remote-access) to get around a similar restriction when he was hosting a minecraft server. Disclaimer: I've never used it so YMMV, just figured it was worth throwing out there

**Reply by J0hnHawk** [reply]
No it's ok for me :-) But I'm not a totaly newbie ;-)

I read the help of `odin notify` and figured out how it could work. This is what works so far in the container:
`odin notify "Aktuelle IP: $(curl -s ipinfo.io/ip)"`

I'll write a script later which will check if the IP is changed (e.g after a disconnect) and message it to Discord if it is changed since the last check.

But unfortunately I've to work now :-/ Thanks for pushing in the right direction.

**Reply by mbround18** [reply]
When I finish the `status` command you could simplify your command by using `jq` to parse the ip address. It will curl your public ip address to spit out in the status block.
